libherokubuildpack/src/command.rs

```diff
@@ -0,0 +1,185 @@
+use std::{io, process, thread};
+use std::{mem, panic};
+
+/// Extension trait for [`process::Command`] that adds functions for use within buildpacks.
+pub trait CommandExt {
+    /// Spawns the command process and sends the output of stdout and stderr to the given writers.
+    ///
+    /// This allows for additional flexibility when dealing with these output streams compared to
+    /// functionality that the stock [`process::Command`] provides. See the [`write`](crate::write)
+    /// module for [`std::io::Write`] implementations designed for common buildpack tasks.
+    ///
+    /// This function will redirect the output unbuffered and in parallel for both streams. This
+    /// means that it can be used to output data from these streams while the command is running,
+    /// providing a live view into the process' output. This function will block until both streams
+    /// have been closed.
+    ///
+    /// # Example:
+    /// ```no_run
+    /// use libherokubuildpack::command::CommandExt;
+    /// use libherokubuildpack::write::tee;
+    /// use std::fs;
+    /// use std::process::Command;
+    ///
+    /// let logfile = fs::File::open("log.txt").unwrap();
+    /// let exit_status = Command::new("date")
+    ///     .spawn_and_write_streams(tee(std::io::stdout(), &logfile), std::io::stderr())
+    ///     .and_then(|mut child| child.wait())
+    ///     .unwrap();
+    /// ```
+    fn spawn_and_write_streams<OW: Write + Send, EW: Write + Send>(
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Child>;
+
+    /// Spawns the command process and sends the output of stdout and stderr to the given writers.
+    ///
+    /// In addition to what [`spawn_and_write_streams`](Self::spawn_and_write_streams) does, this
+    /// function captures stdout and stderr as `Vec<u8>` and returns them after waiting for the
+    /// process to finish. This function is meant as a drop-in replacement for existing
+    /// `Command:output` calls.
+    ///
+    /// # Example:
+    /// ```no_run
+    /// use libherokubuildpack::command::CommandExt;
+    /// use libherokubuildpack::write::tee;
+    /// use std::fs;
+    /// use std::process::Command;
+    ///
+    /// let logfile = fs::File::open("log.txt").unwrap();
+    /// let output = Command::new("date")
+    ///     .output_and_write_streams(tee(std::io::stdout(), &logfile), std::io::stderr())
+    ///     .unwrap();
+    ///
+    /// // Return value can be used as with Command::output, but the streams will also be written to
+    /// // the given writers.
+    /// println!(
+    ///     "Process exited with {}, stdout: {:?}, stderr: {:?}",
+    ///     output.status, output.stdout, output.stderr
+    /// );
+    /// ```
+    fn output_and_write_streams<OW: Write + Send, EW: Write + Send>(
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Output>;
+}
+
+impl CommandExt for process::Command {
```

Comment:
How does this not violate orphan rules? I thought we couldn't implement a trait for a struct we don't own (broad strokes).

Reply:
Since this crate defines the `CommandExt` trait it's fine. The crate implementing a trait must either define the trait _or_ the type it's implemented for.